### PR TITLE
Throw exceptions when pinot grpc query incomplete

### PR DESCRIPTION
**Pull request by INNOCENT-BOY**
bugfix fro trino-pinot connector
We maintain several Pinot Clusters. And some of clients use Trino as query engine due to the limitation of pinot (join or sub-query). And Trino Cluster enabled grpc to increase the speed of fetching data. But in some scenarios, some of clients report some incidents. Although the query is successful, They still couldn't fetch completed query result. We found some error log s about trino grpc that indicates some of servers didn't give the completed result. So we change trino codes to report error when pinot query result comes with exceptions, such timetoutexception.
So it can notify clients query result uncompleted which would drive clients to query again.
Please trino team help me to review.

### CODE REVIEW

File: plugin/trino-pinot/src/main/java/io/trino/plugin/pinot/client/PinotGrpcDataFetcher.java [hunk 1]
```diff
@@ -254,17 +257,24 @@ public Iterator<PinotDataTableWithSize> queryPinot(ConnectorSession session, Str
...
+            return new ResponseIterator(client.submit(serverRequest), query);
         }
 
         public static class ResponseIterator
                 extends AbstractIterator<PinotDataTableWithSize>
         {
             private final Iterator<Server.ServerResponse> responseIterator;
+            private final String query;
 
             public ResponseIterator(Iterator<Server.ServerResponse> responseIterator)
```
Comment (ebyhr) [inline plugin/trino-pinot/src/main/java/io/trino/plugin/pinot/client/PinotGrpcDataFetcher.java:269]: Please remove this unused constructor.

File: plugin/trino-pinot/src/main/java/io/trino/plugin/pinot/client/PinotGrpcDataFetcher.java [hunk 2]
```diff
@@ -254,17 +257,24 @@ public Iterator<PinotDataTableWithSize> queryPinot(ConnectorSession session, Str
...
+            {
+                this(responseIterator, "unknown sql");
+            }
+
+            public ResponseIterator(Iterator<Server.ServerResponse> responseIterator, String query)
             {
                 this.responseIterator = requireNonNull(responseIterator, "responseIterator is null");
+                this.query = query;
```
Comment (ebyhr) [inline plugin/trino-pinot/src/main/java/io/trino/plugin/pinot/client/PinotGrpcDataFetcher.java:277]: Add `requireNonNull`.

File: plugin/trino-pinot/src/main/java/io/trino/plugin/pinot/client/PinotGrpcDataFetcher.java [hunk 3]
```diff
@@ -275,16 +285,22 @@ protected PinotDataTableWithSize computeNext()
                 }
                 Server.ServerResponse response = responseIterator.next();
                 String responseType = response.getMetadataMap().get(MetadataKeys.RESPONSE_TYPE);
-                if (responseType.equals(ResponseType.METADATA)) {
```
Comment (ebyhr) [inline plugin/trino-pinot/src/main/java/io/trino/plugin/pinot/client/PinotGrpcDataFetcher.java:278]: Why did you move this code?

**Review by ebyhr — COMMENTED** [review]
Please fix the commit title. 
https://github.com/trinodb/trino/blob/master/.github/DEVELOPMENT.md#format-git-commit-messages

File: plugin/trino-pinot/src/main/java/io/trino/plugin/pinot/client/PinotGrpcDataFetcher.java (hunk 3 above)
Comment (ebyhr) [reply plugin/trino-pinot/src/main/java/io/trino/plugin/pinot/client/PinotGrpcDataFetcher.java:278]: Reminder.

**Comment by ebyhr** [comment]
@INNOCENT-BOY Please check the above comments.

**Comment by INNOCENT-BOY** [comment]
> https://github.com/trinodb/cla

I have send email to cla，please take your time to review.

**Comment by ebyhr** [comment]
@INNOCENT-BOY Please address the above review comments.

**Comment by INNOCENT-BOY** [comment]
Hi @ebyhr , I have given the response to some your comments and wait to your reply. 
But I am confused about "Reminder" that you said, Could you give me more detail description?

**Comment by ebyhr** [comment]
Please use rebase instead of merge.

File: plugin/trino-pinot/src/main/java/io/trino/plugin/pinot/client/PinotGrpcDataFetcher.java [hunk 4]
```diff
@@ -279,12 +285,18 @@ protected PinotDataTableWithSize computeNext()
                     return endOfData();
                 }
                 ByteBuffer buffer = response.getPayload().asReadOnlyByteBuffer();
+                DataTable dataTable = null;
```
Comment (ebyhr) [inline plugin/trino-pinot/src/main/java/io/trino/plugin/pinot/client/PinotGrpcDataFetcher.java:288]: ` = null` is redundant.

File: plugin/trino-pinot/src/main/java/io/trino/plugin/pinot/client/PinotGrpcDataFetcher.java [hunk 5]
```diff
@@ -279,12 +285,18 @@ protected PinotDataTableWithSize computeNext()
...
                 try {
-                    return new PinotDataTableWithSize(DataTableFactory.getDataTable(buffer), buffer.remaining());
+                    dataTable = DataTableFactory.getDataTable(buffer);
                 }
                 catch (IOException e) {
                     throw new UncheckedIOException(e);
                 }
+                List<String> exceptions = dataTable.getExceptions().entrySet().stream().map(Map.Entry::getValue).collect(toImmutableList());
```
Comment (ebyhr) [inline plugin/trino-pinot/src/main/java/io/trino/plugin/pinot/client/PinotGrpcDataFetcher.java:295]: What's the key of `getExceptions` map? Why do we want to discard it?

File: plugin/trino-pinot/src/main/java/io/trino/plugin/pinot/client/PinotGrpcDataFetcher.java [hunk 6]
```diff
@@ -279,12 +285,21 @@ protected PinotDataTableWithSize computeNext()
...
                 ByteBuffer buffer = response.getPayload().asReadOnlyByteBuffer();
+                DataTable dataTable;
                 try {
-                    return new PinotDataTableWithSize(DataTableFactory.getDataTable(buffer), buffer.remaining());
+                    dataTable = DataTableFactory.getDataTable(buffer);
                 }
                 catch (IOException e) {
                     throw new UncheckedIOException(e);
                 }
+                List<String> exceptions = dataTable.getExceptions().entrySet().stream()
```
Comment (ebyhr) [inline plugin/trino-pinot/src/main/java/io/trino/plugin/pinot/client/PinotGrpcDataFetcher.java:295]: No need to build `exceptions` object when the exception is an empty. Move to under `if`.

File: plugin/trino-pinot/src/main/java/io/trino/plugin/pinot/client/PinotGrpcDataFetcher.java [hunk 7]
```diff
@@ -279,12 +285,21 @@ protected PinotDataTableWithSize computeNext()
...
+                        .map(entry -> String.format("ErrorCode: %d Error Message: %s", entry.getKey(), entry.getValue()))
```
Comment (ebyhr) [inline plugin/trino-pinot/src/main/java/io/trino/plugin/pinot/client/PinotGrpcDataFetcher.java:296]: Static import `format` method.

File: plugin/trino-pinot/src/main/java/io/trino/plugin/pinot/client/PinotGrpcDataFetcher.java [hunk 8]
```diff
@@ -279,12 +285,21 @@ protected PinotDataTableWithSize computeNext()
...
+                        .collect(toImmutableList());
+                if (!exceptions.isEmpty()) {
```
Comment (ebyhr) [inline plugin/trino-pinot/src/main/java/io/trino/plugin/pinot/client/PinotGrpcDataFetcher.java:298]: Change to `if (!dataTable.getExceptions().isEmpty()) {`

File: plugin/trino-pinot/src/main/java/io/trino/plugin/pinot/client/PinotGrpcDataFetcher.java [hunk 9]
```diff
@@ -279,12 +285,21 @@ protected PinotDataTableWithSize computeNext()
...
+                    throw new PinotException(PINOT_EXCEPTION, Optional.of(query),
+                            format("Encountered %d pinot exceptions : %s", exceptions.size(), exceptions));
```
Comment (ebyhr) [inline plugin/trino-pinot/src/main/java/io/trino/plugin/pinot/client/PinotGrpcDataFetcher.java:300]: Unwrap these 2 lines.

File: plugin/trino-pinot/src/main/java/io/trino/plugin/pinot/client/PinotGrpcDataFetcher.java [hunk 10]
```diff
@@ -279,12 +285,26 @@ protected PinotDataTableWithSize computeNext()
...
                 try {
-                    return new PinotDataTableWithSize(DataTableFactory.getDataTable(buffer), buffer.remaining());
+                    dataTable = DataTableFactory.getDataTable(buffer);
                 }
                 catch (IOException e) {
                     throw new UncheckedIOException(e);
                 }
+                if (!dataTable.getExceptions().isEmpty()) {
+                    List<String> exceptions = dataTable.getExceptions().entrySet().stream()
+                            .map(entry -> format("ErrorCode: %d Error Message: %s", entry.getKey(), entry.getValue()))
```
Comment (ebyhr) [inline plugin/trino-pinot/src/main/java/io/trino/plugin/pinot/client/PinotGrpcDataFetcher.java:297]: "ErrorCode" vs "Error Message"

Let's change to "Error code" and "Error message"

File: plugin/trino-pinot/src/main/java/io/trino/plugin/pinot/client/PinotGrpcDataFetcher.java [hunk 11]
```diff
@@ -279,12 +285,26 @@ protected PinotDataTableWithSize computeNext()
...
+                            .collect(toImmutableList());
+
+                    throw new PinotException(PINOT_EXCEPTION,
+                            Optional.of(query),
+                            format("Encountered %d pinot exceptions : %s",
```
Comment (ebyhr) [inline plugin/trino-pinot/src/main/java/io/trino/plugin/pinot/client/PinotGrpcDataFetcher.java:302]: "Encountered %d exceptions: %s"

**Review by ebyhr — APPROVED** [review]

File: plugin/trino-pinot/src/main/java/io/trino/plugin/pinot/client/PinotGrpcDataFetcher.java [hunk 12]
```diff
@@ -279,12 +285,26 @@ protected PinotDataTableWithSize computeNext()
...
+                if (!dataTable.getExceptions().isEmpty()) {
+                    List<String> exceptions = dataTable.getExceptions().entrySet().stream()
+                            .map(entry -> format("Error code: %d Error message: %s", entry.getKey(), entry.getValue()))
+                            .collect(toImmutableList());
+
+                    throw new PinotException(PINOT_EXCEPTION,
+                            Optional.of(query),
+                            format("Encountered %d exceptions: %s",
+                            exceptions.size(),
+                            exceptions));
```
Comment (ebyhr) [inline plugin/trino-pinot/src/main/java/io/trino/plugin/pinot/client/PinotGrpcDataFetcher.java:304]: The indentation isn't correct. Let's make it one line. 
```java
throw new PinotException(PINOT_EXCEPTION, Optional.of(query), format("Encountered %d exceptions: %s", exceptions.size(), exceptions));
```

**Comment by ebyhr** [comment]
Please avoid rebase unless there's a conflict. Otherwise, it's hard to confirm the force pushed diff on UI.
